src/credentials/sso/ssoAccessTokenProvider.ts

```diff
@@ -0,0 +1,148 @@
+                    startUrl: this.ssoUrl,
+                    region: this.ssoRegion,
+                    accessToken: tokenResponse.accessToken!,
+                    expiresAt: new Date(Date.now() + tokenResponse.expiresIn! * 1000).toISOString(),
+                }
+                return accessToken
+            } catch (err) {
+                if (err.code === 'SlowDownException') {
+                    retryInterval += 5000
+                } else if (err.code === 'AuthorizationPendingException') {
+                    if (Date.now() + retryInterval > deviceCodeExpiration) {
+                        throw Error(
+                            `Device code has expired while polling for SSO token, login flow must be re-initiated.`
+                        )
+                    }
+                    // else, wait the interval and try again
+                } else if (err.code === 'ExpiredTokenException') {
+                    throw Error(`Device code has expired while polling for SSO token, login flow must be re-initiated.`)
+                }
+            }
+            setTimeout(() => {}, retryInterval)
+        }
+    }
+
+    private async authorizeClient(registration: ClientRegistration): Promise<Authorization> {
+        const authorizationParams = {
+            clientId: registration.clientId,
+            clientSecret: registration.clientSecret,
+            startUrl: this.ssoUrl,
+        }
+        try {
+            const authorizationResponse = await this.ssoOidcClient
+                .startDeviceAuthorization(authorizationParams)
+                .promise()
+            const authorization: Authorization = {
+                deviceCode: authorizationResponse.deviceCode!,
+                userCode: authorizationResponse.userCode!,
+                verificationUri: authorizationResponse.verificationUri!,
+                verificationUriComplete: authorizationResponse.verificationUriComplete!,
+                expiresIn: authorizationResponse.expiresIn!,
+                interval: authorizationResponse.interval!,
+            }
+
+            const signInInstructionMessage = `You have chosen an AWS Single Sign-On profile that requires authorization. `
+            vscode.window.showInformationMessage(signInInstructionMessage, { modal: true })
+            vscode.env.openExternal(vscode.Uri.parse(authorization.verificationUriComplete))
+            return authorization
+        } catch (err) {
+            throw err
+        }
+    }
+
+    private async registerClient(): Promise<ClientRegistration> {
+        if (this.cache.loadClientRegistration(this.ssoRegion)) {
+            return this.cache.loadClientRegistration(this.ssoRegion)!
+        }
+
+        // If ClientRegistration token is expired or does not exist, register ssoOidc client
+        const registerParams = {
+            clientType: CLIENT_REGISTRATION_TYPE,
+            clientName: `aws-toolkit-vscode-${Date.now()}`,
```

Comment:
client name should be made a constant that is shared with the token cache code (and anywhere else the value is used for this purpose)

Reply:
Changed into a constant, but it's not used anywhere else.